sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -1133,7 +1180,11 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
               Decimal $tmp = Decimal.apply(new java.math.BigDecimal($c.toString().trim()));
               ${changePrecision(tmp, target, evPrim, evNull, canNullSafeCast)}
             } catch (java.lang.NumberFormatException e) {
-              $evNull = true;
+              if ($ansiEnabled) {
```

Comment:
we don't need try-catch at all when ansi=true.

how about
```
val castCode =
  code"""
    Decimal $tmp = Decimal.apply(new java.math.BigDecimal($c.toString().trim()));
    ${changePrecision(tmp, target, evPrim, evNull, canNullSafeCast)}
  """
if (!ansi) {
  code"""
    try {
      $castCode
    } catch ...
  """
} else {
  castCode
}
```

Reply:
Yeah. This is better.

Reply:
@cloud-fan There is one downside of not catching and throwing `NumberFormatException`. Users wont get proper explanation for the exception as we are giving for all the other datatypes.

Reply:
can we update `changePrecision` to throw `NumberFormatException` with better error message?

Reply:
`NumberFormatException` is thrown when we are converting to BigDecimal. So we can update the error message in `changePrecision` but it will be similar to what we are doing here. First we need to catch the exception and process according to state of  `ansi` flag.

Reply:
`changePrecision` always throws exception for overflow, right? We don't need to check the ansi flag there. We only need to look at the ansi flag in `Cast`